code/modules/hydroponics/biogenerator.dm

```diff
@@ -1,337 +1,354 @@
-				for(var/C in categories)
-					if(C in D.category)
-						categories[C] += D
-
-			dat += "<div class='statusDisplay'>Biomass: [points] units.</div><BR>"
-			dat += "<A href='?src=[src.UID()];activate=1'>Activate</A><A href='?src=[src.UID()];detach=1'>Detach Container</A>"
-			for(var/cat in categories)
-				dat += "<h3>[cat]:</h3>"
-				dat += "<div class='statusDisplay'>"
-				for(var/V in categories[cat])
-					var/datum/design/D = V
-					dat += "[D.name]: <A href='?src=[src.UID()];create=[D.UID()];amount=1'>Make</A>"
-					if(cat in timesFiveCategories)
-						dat += "<A href='?src=[src.UID()];create=[D.UID()];amount=5'>x5</A>"
-					if(ispath(D.build_path, /obj/item/stack))
-						dat += "<A href='?src=[src.UID()];create=[D.UID()];amount=10'>x10</A>"
-					dat += "([D.materials[MAT_BIOMASS]/efficiency])<br>"
-				dat += "</div>"
-		else
-			dat += "<div class='statusDisplay'>No container inside, please insert container.</div>"
+		to_chat(user, "<span class='warning'>You cannot put this in [name]!</span>")
 
-	var/datum/browser/popup = new(user, "biogen", name, 350, 520)
-	popup.set_content(dat)
-	popup.open()
-	return
+/**
+ * Builds/Updates the `product_list` used by the UI.
+ */
+/obj/machinery/biogenerator/proc/update_ui_product_list()
+	product_list = list()
+	for(var/category in categories)
+		product_list[category] = list()
+
+	for(var/V in files.known_designs)
+		var/datum/design/D = files.known_designs[V]
+		for(var/category in categories)
+			if(!(category in D.category))
+				continue
+			product_list[category][D.name] = list(
+				"name" = D.name,
+				"id" = D.id,
+				"cost" = D.materials[MAT_BIOMASS]
+			)
 
 /obj/machinery/biogenerator/attack_hand(mob/user)
-	interact(user)
+	ui_interact(user)
 
 /obj/machinery/biogenerator/attack_ghost(mob/user)
-	interact(user)
+	ui_interact(user)
+
+/obj/machinery/biogenerator/ui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = FALSE, datum/tgui/master_ui = null, datum/ui_state/state = GLOB.default_state)
+	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
+	if(!ui)
+		ui = new(user, src, ui_key, "Biogenerator", "Biogenerator", 390, 600, master_ui, state)
+		ui.set_autoupdate(FALSE)
+		ui.open()
+
+/obj/machinery/biogenerator/ui_data(mob/user)
+	var/list/data = list(
+		"processing" = processing,
+		"biomass" = biomass,
+		"efficiency" = efficiency,
+		"has_plants" = length(stored_plants) ? TRUE : FALSE
+	)
+	if(container)
+		data["container"] = TRUE
+		data["container_curr_reagents"] = container.reagents.total_volume
+		data["container_max_reagents"] = container.reagents.maximum_volume
+	else
+		data["container"] = FALSE
+
+	return data
 
-/obj/machinery/biogenerator/proc/activate()
-	if(usr.stat != 0)
+/obj/machinery/biogenerator/ui_static_data(mob/user)
+	return list("product_list" = product_list)
+
+/obj/machinery/biogenerator/ui_act(action, list/params)
+	if(..())
 		return
-	if(stat != 0) //NOPOWER etc
+
+	. = TRUE
```

Comment:
The semantics of `ui_act()` return value scares and confuses me...
I'd think it should be TRUE all the time? As in, even in the `if(..()) return` above?

Or if not, then.. it should be TRUE only when change happens, so `return TRUE` at the end of the proc, and no `. = TRUE` here?
Again, i'm not sure, this is confusing...

Reply:
If the parent `/datum/proc/ui_act` returns false, that means we can't access the UI, and we wouldn't want to set `. = TRUE`, and therefore update the UI before we know that.

As for the `. = TRUE` at the beginning vs `return TRUE` at the end: it's really up to preference here (and tbh I didn't think much about it). Since we know the UI only passes one of the four `action`s, something is going to be hit, and it's going to always be true.